geo/src/main/java/org/apache/carbondata/geo/scan/expression/PolygonRangeListExpression.java

```diff
@@ -0,0 +1,211 @@
+
+  private String polygonRangeList;
+
+  private String opType;
+
+  private List<Long[]> ranges = new ArrayList<Long[]>();
+
+  private ColumnExpression column;
+
+  private static final ExpressionResult trueExpRes =
+      new ExpressionResult(DataTypes.BOOLEAN, true);
+
+  private static final ExpressionResult falseExpRes =
+      new ExpressionResult(DataTypes.BOOLEAN, false);
+
+  public PolygonRangeListExpression(String polygonRangeList, String opType, String columnName) {
+    this.polygonRangeList = polygonRangeList;
+    this.opType = opType;
+    this.column = new ColumnExpression(columnName, DataTypes.LONG);
+  }
+
+  private void processExpression() {
+    try {
+      // 1. parse the range list string
+      List<String> rangeLists = new ArrayList<>();
+      Pattern pattern = Pattern.compile(GeoConstants.RANGELIST_REG_EXPRESSION);
+      Matcher matcher = pattern.matcher(polygonRangeList);
+      while (matcher.find()) {
+        String matchedStr = matcher.group();
+        rangeLists.add(matchedStr);
+      }
+      // 2. process the range lists
+      if (rangeLists.size() > 0) {
+        List<Long[]> processedRangeList = getRangeListFromString(rangeLists.get(0));
+        for (int i = 1; i < rangeLists.size(); i++) {
+          List<Long[]> tempRangeList = getRangeListFromString(rangeLists.get(i));
+          processedRangeList = GeoHashUtils.processRangeList(
+            processedRangeList, tempRangeList, opType);
+        }
+        ranges = processedRangeList;
+        GeoHashUtils.validateRangeList(ranges);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void sortRange(List<Long[]> rangeList) {
+    rangeList.sort(new Comparator<Long[]>() {
+      @Override
+      public int compare(Long[] x, Long[] y) {
+        return Long.compare(x[0], y[0]);
+      }
+    });
+  }
+
+  private void combineRange(List<Long[]> rangeList) {
+    if (rangeList.size() > 1) {
+      for (int i = 0, j = i + 1; i < rangeList.size() - 1; i++, j++) {
+        long previousEnd = rangeList.get(i)[1];
+        long nextStart = rangeList.get(j)[0];
+        if (previousEnd + 1 == nextStart) {
+          rangeList.get(j)[0] = rangeList.get(i)[0];
+          rangeList.get(i)[0] = null;
+          rangeList.get(i)[1] = null;
+        }
+      }
+      rangeList.removeIf(item -> item[0] == null && item[1] == null);
+    }
+  }
+
+  private List<Long[]> getRangeListFromString(String rangeListString) {
+    String[] rangeStringList = rangeListString.trim().split(GeoConstants.DEFAULT_DELIMITER);
+    List<Long[]> rangeList = new ArrayList<>();
+    for (String rangeString : rangeStringList) {
+      String[] range = GeoHashUtils.splitStringToPoint(rangeString);
+      if (range.length != 2) {
+        throw new RuntimeException("each range is a pair need 2 data");
+      }
+      try {
+        rangeList.add(new Long[] {Long.valueOf(range[0]), Long.valueOf(range[1])});
```

Comment:
Better check if Long.valueOf(range[0]) < Long.valueOf(range[1]) since it is user input.

Reply:
Added this checking